Change Consoles data fetching time - fetch Consoles data on demand instead of fetch on login

bamsalem opened the pull request:
This PR should test with engine that include this [gerrit patch](https://gerrit.ovirt.org/#/c/ovirt-engine/+/112536/) and will need to be merged after the patch is merged.
depended on gerrit patch https://gerrit.ovirt.org/112536.

Resolves: https://bugzilla.redhat.com/show_bug.cgi?id=1795457

This change include 4 sub changes:
1. Removes the graphic console from `getVms` request in case of `shallowFetch`, that action should reduce the CPU performance issue of the postgreSQL server. this change is in commit 4e52833 .
2. Adds actions, reducers and sagas.
Those changes allow fetch the consoles separately and included in commit 4e52833.
3. Adds new config value `defaultConsole` to the app config, this value is the system level `defaultConsole` in the future we can add more levels like `userDefaultConsole`, `vmDefaultConsole`, etc..
Those changes are in f60363c .
4. Adds new options to `vmActions` when it is on card 'Console' .
When User clicks on the 'Console' button 2 things happens: app makes an api call that fetches consoles data,
and when the call is finished it selects automatically console from the consoles list.


Error handling and few assumptions:
1. If it cannot select console from the list the app will use `ToastNotifications` to notify the user
2. If The list is empty it will notify the user.
3. In case The list includes only one protocol it will select this one automatically.
4. The default console is 'vnc' since 'spice' won't be supported from RHEL 8.3 .
5. In case of windows it won't select the RDP as the default console it will select the spice.
6. RDP is always shown in the `vmActions` list since it's not require any data from the consoles data.

Screenshots:
Before:
![image](https://user-images.githubusercontent.com/64131213/101489637-e5854d80-3969-11eb-9c8a-5edeae2c5f69.png)

After:
![image](https://user-images.githubusercontent.com/64131213/101489528-bec71700-3969-11eb-81b6-1654b83b59ca.png)

ovirt-infra commented:
Hello contributor, thanks for submitting a PR for this project! 

I am the bot who triggers "standard-CI" builds for this project.
As a security measure, I will not run automated tests on PRs that are not from white-listed contributors.

In order to allow automated tests to run, please ask one of the project maintainers to review the code and then do one of the following:

1. Type `ci test please` on this PR to trigger automated tests for it.
2. Type `ci add to whitelist` on this PR to trigger automated tests for it and also add you to the contributor white-list so that your future PRs will be tested automatically. ( keep in mind this list might be overwritten if the job XML is refreshed, for permanent whitelisting, please follow #3 option )
3. If you are planning to contribute to more than one project, maybe it's better to ask them to add you to the project organization, so you'll be able to run tests for all the organization's projects.

sjd78 commented:
ci add to whitelist

rszwajko commented on src/sagas/index.js:
```diff
@@ -183,6 +184,16 @@ export function* fetchByPage () {
   }))
 }
 
+export function* fetchConsoles ({ payload: { vmId } }) {
```
Exporting this function seems not necessary. Consider also removing `return` as this saga is not called directly.

rszwajko commented on src/components/VmActions/index.js:
```diff
@@ -134,15 +136,40 @@ class VmActions extends React.Component {
-      ),
-    })).toJS()
+    let consoles = []
+    if (isOnCard) {
```
`isCard` was already used before to separate card-scenario from toolbar scenario - however the `getDeafaultAction` step was common to both paths. Right know we no longer have a clean point where the scenarios diverge. Perhaps it's time to break this method into smaller ones i.e. `getDefaultActionsForCard()` and `getDefaultActionsForToolbar()` ?

bamsalem replied:
IMHO most of the logic is shared and both of the options using it, I think that we should not separate it,
but if most of the team thinks different I'll change it.

sgratch replied:
@bamsalem @rszwajko I don't have a strong opinion on this. The only difference is the consoles menu so not sure it worth separating in current phase.

rszwajko replied:
ok. let's leave it as it is.

bamsalem commented on src/components/VmActions/AutoSelectConsole.js:
```diff
@@ -0,0 +1,112 @@
+import ConsoleConfirmationModal from '_/components/VmActions/ConsoleConfirmationModal'
+import { msg } from '_/intl'
+
+class AutoSelectConsole extends Component {
```
> 1. it's a new code so I would stick to the functional component pattern

Done.
> 2. `setTimeout` usually requires some `clearTimeout` or we risk triggering action on unmounted component

we don't use `setTimeout` anymore
> 3. I've seen following error when clicking Console button
> 
> > Warning: Cannot update during an existing state transition (such as within `render`). Render methods should be a pure function of props and state.

`onFailure` is updating the state, I moved the call out from `render()`

sgratch commented on src/intl/messages.js:
```diff
@@ -61,7 +61,12 @@ export const messages: { [messageId: string]: MessageType } = {
   cancel: 'Cancel',
   cannotUpdateCloudInitHostname: 'Host name cannot be synchronized with VM name because the VM name is not a valid host name.',
+  cantDownloadConsoleEmptyList: 'There are no consoles assigned to this VM.',
+  cantDownloadConsoleGeneral: "Can't download console:",
```
Please rephrase to something more user friendly like: "can't open a graphical console since this is  a headless VM"
Please also add the VM name to the message.

rszwajko replied:
> "non RDP Console"...

I would rather not name a key functionality using Windows protocol. It looks as RDP was the default and other consoles were our second choice.

> the RDP seems to be neglected as a separate option, not related to other consoles which is confusing

@bamsalem  Can we treat RDP the same as other consoles? This would remove the source of confusion.

sgratch replied:
@rszwajko 

> > "non RDP Console"...
> I would rather not name a key functionality using Windows protocol. It looks as RDP was the default and other consoles were our second choice.
> 
> > the RDP seems to be neglected as a separate option, not related to other consoles which is confusing
> Can we treat RDP the same as other consoles? This would remove the source of confusion.

I agree. I suggested calling this: "spice/vnc console", "rdp console"

The problem with treating RDP the same as other consoles is that RDP is always enabled for windows based OS vms only, regardless to spice/vnc. So we'll need to add another config value for setting global default console for windows based vms and one for non windows based vms instead of one "ClientModeConsoleDefault". This will be required for account settings solution but I wanted to save that extra config parameter. If you think it's better then let's add this.

sgratch reviewed (dismissed):
Few comments:

- You didn't read the value of the ClientModeVncDefault config val for deciding which VNC mode to use (native or noVNC) as we already discussed.
- In case of a Windows OS, the RDP seems to be neglected as a separate option, not related to other consoles which is confusing. e.g. in case of headless the "no consoles assigned" message is displayed even though there is the RDP option. I suggest renaming the "console" pull-down menu option in case of Windows base to something else and the error message as well - maybe to "non RDP Console"...
- Any reason why not removing the follow=graphics_consoles for manual refresh as well? We are anyway not using the graphics console info after manual refresh on vms list, right?
- After generating the "Can't download console: There are no consoles assigned to this VM." error when trying to open a headless VM console, this notification message re-appear and added to drawer once per each auto refresh. You need to somehow clear this.

rszwajko commented on src/components/VmActions/AutoSelectConsole.js:
```diff
@@ -0,0 +1,133 @@
+        failureReason = 'emptyConsolesList'
+      }
+    }
+    return { selectedProtocol, failureReason }
```
Function can be made pure by moving `defaultConsole` to parameters. We could then move it outside the component and unit test. This would also allow to use it directly as hook dependency.

bamsalem replied:
Done.

rszwajko commented on src/components/VmActions/AutoSelectConsole.js:
```diff
@@ -0,0 +1,133 @@
+  } = state
+
+  useEffect(() => {
+    if (show) {
```
Consider data driven design instead of introducing state flags (`active, dispatchStatus, show, ...`).
Basic flow:
1. No data -> fetch data (to the Redux store)
2. there is data -> render data

Check [VmCardList](https://github.com/oVirt/ovirt-web-ui/blob/8a5a26cb2c3310a67d05673d3482bf1a751cc0a9/src/components/VmsList/VmCardList.js#L44)

bamsalem replied:
@rszwajko  I'm not sure if it's the same case, in `vmCardList` data fetching should be triggered automatically, but in this case it should be triggered by click, so we must use flags to avoid fetching when it's not needed...

rszwajko replied:
Usually in React/Redux a component just renders data from props. If we need more data then we dispatch an action that fetches missing data and saves it in the store. From there the components are notified via props.
Fetch logic (i.e. error handling) is placed in saga-layer.

In case of AutoSelect the rendered piece of data is `selectedProtocol` - we could place it in the store and populate via some kind of `fetchAndSelectConsole` saga. This would remove all error/state handling logic from the component. Perhaps it would even make the component obsolete since it's more a date provider then a regular component.

rszwajko commented on src/components/VmActions/AutoSelectConsole.js:
```diff
@@ -0,0 +1,133 @@
+    setState(state => ({ ...state, ...newState }))
+  }, [vm, requestActive]) // eslint-disable-line react-hooks/exhaustive-deps
+
+  const onClose = () => {
```
why we need `onClose` here?

bamsalem replied:
To update the state before call the `props.onClose` function, but it's changed so take a look on the latest version :)

rszwajko commented on src/components/VmActions/AutoSelectConsole.js:
```diff
@@ -0,0 +1,133 @@
+    defaultConsole: state.config.get('defaultConsole'),
+    requestActive: !state.activeRequests.isEmpty(),
+  }),
+  (dispatch, { defaultConsole }) => ({
```
`defaultConsole` here is passed as `ownProps` but is retrieved from the store. Usually `ownProps` and store based are different categories.

rszwajko commented on src/components/VmActions/AutoSelectConsole.js:
```diff
@@ -0,0 +1,133 @@
+      const generalMsg = msg.cantDownloadConsoleGeneral()
+      switch (reason) {
+        case 'emptyConsolesList':
+          dispatch(addUserMessage({ message: `${generalMsg} ${msg.cantDownloadConsoleEmptyList()}`, type: 'error' }))
```
I'd rather avoid concatenating messages in the code as this prevents proper internationalization (grammar differences, write order etc)

bamsalem commented on src/components/VmActions/AutoSelectConsole.js:
```diff
@@ -0,0 +1,133 @@
+      }))
+      fetchConsoles(vm.get('id'))
+    }
+  }, [show]) // eslint-disable-line react-hooks/exhaustive-deps
```
IMO the vm, the fetchConsole shouldn't trigger the function since nothing happened unless show === true.
so why it should be in the dependency list? vm might change on any refresh but it shouldn't affect the AutoSelectConsole component. WDYT?

rszwajko replied:
OK. The we should provide VM ID as dependency(i.e. via props or local variable).
```....
    fetchConsoles(vmId)}}, [show, vmId])
```

rszwajko commented on src/components/VmActions/index.js:
```diff
@@ -134,15 +136,40 @@ class VmActions extends React.Component {
+          actionDisabled: config.get('websocket') === null,
+          id: `${idPrefix}-button-console-browser`,
+          confirmation: (
+            <ConsoleConfirmationModal isNoVNC consoleId={vncConsole.get('id')} vm={vm} />
```
Using `isNoVNC=true` for VNC is confusing. Perhaps we should re-name?
Additionally I think the intention here was probably to pass `isNoVnc` property not just `true`.
See [JSX syntax](https://reactjs.org/docs/jsx-in-depth.html#props-default-to-true)

bamsalem replied:
It should be removed, I forgot to remove it:
>You didn't read the value of the ClientModeVncDefault config val for deciding which VNC mode to use (native or noVNC) as we already discussed.

Thanks

bamsalem commented:
@sgratch 
> Few comments:
> 
> * You didn't read the value of the ClientModeVncDefault config val for deciding which VNC mode to use (native or noVNC) as we already discussed.

Done.
> * In case of a Windows OS, the RDP seems to be neglected as a separate option, not related to other consoles which is confusing. e.g. in case of headless the "no consoles assigned" message is displayed even though there is the RDP option. I suggest renaming the "console" pull-down menu option in case of Windows base to something else and the error message as well - maybe to "non RDP Console"...

Done: 
![image](https://user-images.githubusercontent.com/64131213/104119787-b5e7ad80-533a-11eb-9914-27d60c28d5ea.png)

> * Any reason why not removing the follow=graphics_consoles for manual refresh as well? We are anyway not using the graphics console info after manual refresh on vms list, right?

No reason, removed.
> * After generating the "Can't download console: There are no consoles assigned to this VM." error when trying to open a headless VM console, this notification message re-appear and added to drawer once per each auto refresh. You need to somehow clear this.

Thanks, I didn't notice that ...
It's been handled.

bamsalem commented on src/components/VmActions/index.js:
```diff
@@ -134,15 +136,40 @@ class VmActions extends React.Component {
+        priority: 1,
+        actionDisabled: isPool || !canConsole(status) || vm.getIn(['actionInProgress', 'getConsole']),
+        shortTitle: msg.console(),
+        onClick: (e) => { e.preventDefault() },
```
RDP will not launch via auto select, it's a separate option:
![image](https://user-images.githubusercontent.com/64131213/104282014-113eaa80-54b7-11eb-823b-a698ba4f6d55.png)

bamsalem commented on src/components/VmActions/AutoSelectConsole.js:
```diff
@@ -0,0 +1,122 @@
+export default connect(
+  (state) => ({
+    defaultConsole: state.config.get('defaultConsole'),
+    requestActive: !state.activeRequests.isEmpty(),
```
It's something that I took from `VmList`, we can make it more specific (pseudo)
```
    requestActive: !!state.activeRequests.find(r => /^\/api\/vms\/([0-9]|[A-Z]|[a-z]|-)*\/graphicsconsoles$/.test(r)),
```
WDYT?

rszwajko replied:
It fixes the `any active request` part but it is still a hack. I would rather remove it unless there is really no other way.

rszwajko commented:
@sgratch  @sjd78 
we already have a slice in Redux store dedicated to console [options](https://github.com/oVirt/ovirt-web-ui/blob/8a5a26cb2c3310a67d05673d3482bf1a751cc0a9/src/sagas/console/index.js#L126). Until now it was used exclusively by auto-connect feature which is currently dead (but will probably get resurrected at some point). Auto-selecting protocol seems to fit there.

rszwajko commented on src/components/VmActions/AutoSelectConsole.js:
```diff
@@ -0,0 +1,144 @@
+      }
+      setState(state => ({ ...state, ...newState }))
+    }
+  }, [vm, dispatchStatus, show])
```
eslint warning about dependencies

rszwajko commented on src/components/VmActions/AutoSelectConsole.js:
```diff
@@ -0,0 +1,144 @@
+  useEffect(() => {
+    if (show && dispatchStatus === SHOULD_DISPATCH) {
+      const newState = {}
+      const selectedProtocol = vm.get('defaultConsole')
```
`defaultConsole` is used here in 2 contexts: `config.defaultConsole` and `vm.defaultConsole`.
Perhaps we should rename the second (according to its usage) to `selectedProtocol`?

bamsalem replied:
IMO use `selectedProtocol` can be confusing because inside `AutoSelectConsole` component it's the selectedConsole but in the context of the vm is the default, it's individual and some people might think different...

I leave it that way unless the rest of the team thinks different...

rszwajko commented on src/components/VmActions/index.js:
```diff
@@ -134,15 +136,40 @@ class VmActions extends React.Component {
+          actionDisabled: config.get('websocket') === null,
+          id: `${idPrefix}-button-console-browser`,
+          confirmation: (
+            <ConsoleConfirmationModal isNoVNC consoleId={vncConsole.get('id')} vm={vm} />
```
It's still there:)

sjd78 replied:
`isNoVNC` is the correct attribute to be there in this case.  It tells the confirmation modal to open VNC in the "noVnc" in-browser mode.  The map through the consoles above will generate the confirmation modal to open VNC in the "console.vv" style.

rszwajko replied:
@sjd78  @bamsalem 
OK.  I was confused by the `isNoVnc` prop introduced in AutoSelectConsole which was very similar to `isNoVNC` prop used in ConsoleConfirmationModal. Not passing this property to  ConsoleConfirmationModal seemed a typo to me.
I still think that launching VNC console with `isNoVNC == true` is confusing. Perhaps a different name would be better explain the purpose? However this refactoring is out of scope of this patch so let's leave it as it is.

rszwajko reviewed (changes requested):
@sgratch @sjd78 @hstastna  Take a look at the topics below.  I would like to hear your opinion about the direction we should take:
1.  [isCard ](https://github.com/oVirt/ovirt-web-ui/pull/1342#discussion_r545258589)
2.  [RDP](https://github.com/oVirt/ovirt-web-ui/pull/1342#discussion_r555746901)
3.  [console slice](https://github.com/oVirt/ovirt-web-ui/pull/1342#issuecomment-758660205)

rszwajko commented on src/components/VmActions/AutoSelectConsole.js:
```diff
@@ -0,0 +1,111 @@
+      } else {
+        const consoles = vm.get('consoles')
+        setState({
+          selectedProtocol: consoles.find(console => console.get('protocol') === selectedProtocol),
```
This seem to duplicate logic from `selectDefaultConsoleProtocol()`. Actually it seems to me that `vm.get('defaultConsole')` could be used directly as `props.selectedProtocol` without the need of storing it in local state (own props already contain `vm` so we can retrieve it in `connect()`).

rszwajko commented on src/components/VmActions/AutoSelectConsole.js:
```diff
@@ -0,0 +1,111 @@
+    }
+  }, [vm, dispatchStatus, show, handleClose])
+
+  return (
```
Can we return empty component when `!show || !selectedProtocol`? This would simplify the logic here - probably the dispatch status could be dropped.

sgratch commented:
> You didn't read the value of the ClientModeVncDefault config val for deciding which VNC mode to use (native or noVNC) as we already discussed.

It still doesn't work to me. The noVnc console is not activated.

>     In case of a Windows OS, the RDP seems to be neglected as a separate option, not related to other consoles which is confusing. e.g. in case of headless the "no consoles assigned" message is displayed even though there is the RDP option. I suggest renaming the "console" pull-down menu option in case of Windows base to something else and the error message as well - maybe to "non RDP Console"...

This looks a bit strange to display those 2 options of no rdp and rdp: 
![image](https://user-images.githubusercontent.com/18169498/106031099-82957480-60d7-11eb-9579-5fe3370a665e.png)

In addition, in case of a headless vm then no console should be supported including RDP.

>     After generating the "Can't download console: There are no consoles assigned to this VM." error when trying to open a headless VM console, this notification message re-appear and added to drawer once per each auto refresh. You need to somehow clear this.

Now there is no error at all.

sgratch commented:
> You didn't read the value of the ClientModeVncDefault config val for deciding which VNC mode to use (native or noVNC) as we already discussed.

It still doesn't work to me. The noVnc console is not activated.

>     In case of a Windows OS, the RDP seems to be neglected as a separate option, not related to other consoles which is confusing. e.g. in case of headless the "no consoles assigned" message is displayed even though there is the RDP option. I suggest renaming the "console" pull-down menu option in case of Windows base to something else and the error message as well - maybe to "non RDP Console"...

This looks a bit strange to display those 2 options of non rdp and rdp:
![image](https://user-images.githubusercontent.com/18169498/106031099-82957480-60d7-11eb-9579-5fe3370a665e.png)
Maybe chage to: spice/vnc console, rdp console

In addition, in case of a headless vm then no console should be supported including RDP.

>     After generating the "Can't download console: There are no consoles assigned to this VM." error when trying to open a headless VM console, this notification message re-appear and added to drawer once per each auto refresh. You need to somehow clear this.

Now there is no error at all.

sgratch commented:
> @sgratch @sjd78
> we already have a slice in Redux store dedicated to console [options](https://github.com/oVirt/ovirt-web-ui/blob/8a5a26cb2c3310a67d05673d3482bf1a751cc0a9/src/sagas/console/index.js#L126). Until now it was used exclusively by auto-connect feature which is currently dead (but will probably get resurrected at some point). Auto-selecting protocol seems to fit there.

This might be used when the real "console option settings" as part of account settings feature will be implemented per user with a backend persistently. For now we implement a very basic flow in which for all users we use a backend global settings for console options by querying the config vals.

sgratch commented:
@rszwajko 
>Take a look at the topics below. I would like to hear your opinion about the direction we should take
>     1. [isCard ](https://github.com/oVirt/ovirt-web-ui/pull/1342#discussion_r545258589)
>     2. [RDP](https://github.com/oVirt/ovirt-web-ui/pull/1342#discussion_r555746901)
>     3. [console slice](https://github.com/oVirt/ovirt-web-ui/pull/1342#issuecomment-758660205)

I answered all 3 topics above by replying each separately.

bamsalem commented:
@sgratch @rszwajko  
I made all the change you requested:
Sharon regarding headless VM:
Windows headless vm can use RDP in webAdmin.
The default for windows vm before these change is spice.
Regarding non RDP VM, I change it to "Console" and "Remote Desktop"
![image](https://user-images.githubusercontent.com/64131213/106569495-45007380-653d-11eb-93da-a5c47fef706e.png)

sgratch commented:
@bamsalem I didn't verify all  yet but found that 2 issue mentioned above https://github.com/oVirt/ovirt-web-ui/pull/1342#issuecomment-768462866 are still reproduced:

- The noVnc console is not activated.
   use case: 
   1. set ClientModeConsoleDefault=vnc, ClientModeVncDefault=NoVnc
   2. create a VM with vnc supported
   3. open console for the VM
   4. vnc console is activated instead of  noVnc

- In addition, in case of a headless vm, the console option appears as enabled but nothing happens when clicking on it - no error appears.

bamsalem commented:
> @bamsalem I didn't verify all yet but found that 2 issue mentioned above [#1342 (comment)](https://github.com/oVirt/ovirt-web-ui/pull/1342#issuecomment-768462866) are still reproduced:
> 
> * The noVnc console is not activated.
>   use case:
>   
>   1. set ClientModeConsoleDefault=vnc, ClientModeVncDefault=NoVnc
>   2. create a VM with vnc supported
>   3. open console for the VM
>   4. vnc console is activated instead of  noVnc
> * In addition, in case of a headless vm, the console option appears as enabled but nothing happens when clicking on it - no error appears.

@sgratch 
Some of my changes wasn't staged, please Review it again

sgratch commented on src/actions/index.js:
```diff
@@ -25,6 +25,8 @@ import {
   STOP_SCHEDULER_FOR_RESUMING_NOTIFICATIONS,
   UPDATE_PAGING_DATA,
+  SET_DEFAULT_CONSOLE,
+  SET_VNC_MODE,
```
Since we read those from config and we know that we'll probably override this by user settings in the future, I suggest to rename default_console and vnc_mode on all places in code to something like global-default-console and global-default-vnc-mode accordingly.

sgratch commented on src/actions/vm.js:
```diff
@@ -307,6 +306,17 @@ export function updateIcons ({ icons }) {
+    payload: {
+      vmId,
+      consoles: consolesList,
+      selectedProtocol,
```
Again regarding naming:
Sometimes we use protocol and sometimes console which actually it's the same meaning in our code. So we need to be more persistent.
For example, instead of selectedProtocol here, we can have selectedConsole (on all places in code).

sgratch commented on src/intl/messages.js:
```diff
@@ -62,6 +62,13 @@ export const messages: { [messageId: string]: MessageType } = {
   cancel: 'Cancel',
   cannotUpdateCloudInitHostname: 'Host name cannot be synchronized with VM name because the VM name is not a valid host name.',
   cantLogonToConsole: 'Single sign on failed. Please check to see if the guest agent is running on your virtual machine. Contact your administrator if the problem continues.',
+  cantOpenConsole: "Can't open console: {message}",
```
when should this message appear? it's not informative and seems redundant.

bamsalem replied:
![image](https://user-images.githubusercontent.com/64131213/109476036-1bc0fd80-7a7f-11eb-9e30-655d47af3613.png)

sgratch commented on src/intl/messages.js:
```diff
@@ -62,6 +62,13 @@ export const messages: { [messageId: string]: MessageType } = {
   cannotUpdateCloudInitHostname: 'Host name cannot be synchronized with VM name because the VM name is not a valid host name.',
   cantLogonToConsole: 'Single sign on failed. Please check to see if the guest agent is running on your virtual machine. Contact your administrator if the problem continues.',
+  cantOpenConsole: "Can't open console: {message}",
+  cantOpenGraphicalConsoleEmptyList: "Can't open graphical console since {vmName} is a headless VM.",
```
please rephrase/shorten to "Can't open a console since VM (vmname} is headless."

sgratch commented on src/intl/messages.js:
```diff
@@ -62,6 +62,13 @@ export const messages: { [messageId: string]: MessageType } = {
   cantLogonToConsole: 'Single sign on failed. Please check to see if the guest agent is running on your virtual machine. Contact your administrator if the problem continues.',
+  cantOpenConsole: "Can't open console: {message}",
+  cantOpenGraphicalConsoleEmptyList: "Can't open graphical console since {vmName} is a headless VM.",
+  cantOpenGraphicalConsoleMoreOptions: "Can't open graphical console you can find more options in the VM page.",
```
when should this message appear? it's not informative and seems redundant.

bamsalem replied:
It's a fallback message with the latest changes is redundant  I removed that :)

sgratch commented on src/intl/messages.js:
```diff
@@ -62,6 +62,13 @@ export const messages: { [messageId: string]: MessageType } = {
+  cantOpenConsole: "Can't open console: {message}",
+  cantOpenGraphicalConsoleEmptyList: "Can't open graphical console since {vmName} is a headless VM.",
+  cantOpenGraphicalConsoleMoreOptions: "Can't open graphical console you can find more options in the VM page.",
+  cantOpenGraphicalConsoleNoDefault: "Can't open graphical console {vmName} has no {defaultConsole} console assigned. You can find more options in the VM page.",
```
Please rephrase to:
Can't open a console since VM {vmName} is not assigned with the default console {defaultConsole}....

sgratch commented on src/intl/messages.js:
```diff
@@ -62,6 +62,13 @@ export const messages: { [messageId: string]: MessageType } = {
+  cantOpenGraphicalConsoleNoDefault: "Can't open graphical console {vmName} has no {defaultConsole} console assigned. You can find more options in the VM page.",
+  cantOpenNonRdpGraphicalConsoleEmptyList: "Can't open non Rdp graphical console since {vmName} is a headless VM.",
+  cantOpenNonRdpGraphicalConsoleMoreOptions: "Can't open non Rdp graphical console you can find more options in the VM page.",
+  cantOpenNonRdpGraphicalConsoleNoDefault: "Can't open non Rdp graphical console {vmName} has no {defaultConsole} console assigned. You can find more options in the VM page.",
```
relevant to all new added console messages:
1. "You can find more options in the VM page" - what do you mean by that? Which info on which page?
2.Why do we need those "non RDP" message? Console is the same regardless to RDP.

bamsalem replied:
>1. "You can find more options in the VM page" - what do you mean by that? Which info on which page?

I meant to inform users that in case of no default they can find more options like: RDP, another console in the vm details page.
>2. Why do we need those "non RDP" messages? Console is the same regardless to RDP.

It's just mention that RDP is available, I'll remove that...

sgratch commented on src/sagas/index.js:
```diff
@@ -109,9 +122,7 @@ const VM_FETCH_ADDITIONAL_DEEP = [
-const VM_FETCH_ADDITIONAL_SHALLOW = [
-  'graphics_consoles',
-]
+const VM_FETCH_ADDITIONAL_SHALLOW = []
```
Do we still need this VM_FETCH_ADDITIONAL_SHALLOW?

sgratch commented on src/utils/index.js:
```diff
@@ -40,3 +42,25 @@ export function dividers (num: number): Array<number> {
   return divs
 }
+
+export function getNoDefaultConsoleErrorMessages (reason: string, vmName: string, hasRdp: boolean, defaultConsole: string): string {
```
Why do we need  `hasRdp` for distinguishing between non rdp and console on current solution?  Ithink we should have one message referring to console only.

sgratch commented on src/utils/index.js:
```diff
@@ -40,3 +42,25 @@ export function dividers (num: number): Array<number> {
+        ? msg.cantOpenNonRdpGraphicalConsoleNoDefault({ vmName, defaultConsole })
+        : msg.cantOpenGraphicalConsoleNoDefault({ vmName, defaultConsole })
+      break
+    default:
```
It seems to me that this default case is never reached. Can you explain when we'll get this error and why?

bamsalem replied:
> I'd rather the messages be selected in the place they're needed than be in this util function...

I did it thanks :)

sgratch reviewed (changes requested):
@bamsalem 

- In case of a headless VM, RDP shouldn't be available as well - this is the behavior on webadmin and this was the behaviour on current web-ui on both vms dashboard (prior to this fix) and vm details dialogs.

- Please also update first comment by adding a dependency on patch https://gerrit.ovirt.org/112536.

- Check inline comments.

sjd78 reviewed (changes requested):
A few gaps to close.  Most of my comments are inline.

A thought about the PR's first comment -- you can keep going back and editing it to match the current state of the PR.  For the work you've done here, I'd recommend an overview of the functional changes made and then as much about the technical as makes sense.

E.g. - This PR changes console handling on the Card List page.  Due to the performance issues detailed in BZ123, the graphics_consoles are now fetched asynchronously.  A single "Console" button on each VM Card will use system config values to auto select what console to display to the user.  (Detail how the auto-select works with the config values later).  The Console buttons on the VM details page remain unchanged.

sjd78 commented:
@rszwajko ...

> @sgratch @sjd78 @hstastna Take a look at the topics below. I would like to hear your opinion about the direction we should take:
> 
>    1. [isCard ](https://github.com/oVirt/ovirt-web-ui/pull/1342#discussion_r545258589)

I'd rather refactor the VmActions all together and have a real split and distinct components between 1. On Card, 2. On Details but narrow width / kebab, 3. On Details but normal width / button toolbar.  They can still have the action handlers and confirmation popups, but the blended renders really really annoy me.

> 
>    2. [RDP](https://github.com/oVirt/ovirt-web-ui/pull/1342#discussion_r555746901)
> 

Consoles for a VM are either Spice or VNC (native) or VNC (web).  RDP isn't really a console (even when you tick the connect to console session box in the Remote Desktop app on windows) since it doesn't even become available until the system is running.  You can't RDP to a windows installer etc.  Having the distinct button on the card action drop down or the details page toolbar console drop down is sufficient IMHO.

>    3. [console slice](https://github.com/oVirt/ovirt-web-ui/pull/1342#issuecomment-758660205)

IIRC the slice is more for handling SSO and console access privlidges than anything else.  Just go ahead and try opening a console from and admin user, and then try to open the console for a normal user.  For fun you can get make sure SSO via guest agent is enabled on your VM and see those dialogs (once per app session at least).

bamsalem commented:
@sgratch 

> @bamsalem
> 
> * In case of a headless VM, RDP shouldn't be available as well - this is the behavior on webadmin and this was the behaviour on current web-ui on both vms dashboard (prior to this fix) and vm details dialogs.

I tested that on brq-dev, and it seems like in case of windows  headless VM RDP is still available:
![image](https://user-images.githubusercontent.com/64131213/107874908-9ad20580-6ec5-11eb-86f9-5f813b9722a1.png)
![headlessRdp-2021-02-14_13 15 00](https://user-images.githubusercontent.com/64131213/107875116-b984cc00-6ec6-11eb-8d9a-123ab0227271.gif)

So I'm not sure if it should be available or it's a bug...

bamsalem commented on src/reducers/utils.js:
```diff
@@ -66,3 +70,14 @@ export function removeMissingItems ({ state, subStateName, idsToPreserve }) {
     .asImmutable()
   return state.set(subStateName, newItems)
 }
+export function selectDefaultConsoleProtocol (defaultProtocol, consoles = []) {
```
I'm not sure if it's better than notify user that the default is not available and then he can use VM details and select one of the options...

sgratch commented:
@bamsalem 
> I tested that on brq-dev, and it seems like in case of windows headless VM RDP is still available

Correct, it is enabled on webadmin (my mistake, sorry) and disabled on current web-ui for headless. 
The correct behaviour is to enable RDP for web-ui as well since no graphic protocol is needed to be installed on that VM so headless Windows VMs can be managed using RDP or SSH access and that's the right behaviour.

sgratch commented on src/constants/index.js:
```diff
@@ -120,7 +120,10 @@ export const SET_VM_SESSIONS = 'SET_VM_SESSIONS'
+export const SET_VM_CONSOLES = 'SET_VM_CONSOLES'
 export const SET_WEBSOCKET = 'SET_WEBSOCKET'
+export const SET_GLOBAL_DEFAULT_CONSOLE = 'SET_GLOBAL_DEFAULT_CONSOLE'
+export const SET_GLOBAL_VNC_MODE = 'SET_GLOBAL_VNC_MODE'
```
SET_GLOBAL_DEFAULT_VNC_MODE

sgratch commented on src/intl/messages.js:
```diff
@@ -62,6 +62,7 @@ export const messages: { [messageId: string]: MessageType } = {
   cancel: 'Cancel',
   cannotUpdateCloudInitHostname: 'Host name cannot be synchronized with VM name because the VM name is not a valid host name.',
   cantLogonToConsole: 'Single sign on failed. Please check to see if the guest agent is running on your virtual machine. Contact your administrator if the problem continues.',
+  cantOpenConsole: 'Can\'t open console: {message}',
```
I still don't understand why we added this message (Based on Scott's question: "why did we touch the the SSO code that would warrant adding an error toast message?") and anyway it's not informative enough.

bamsalem replied:
I thought it will be better if we will notify users if any error occurs during console download...
If it not necessary I'll remove that.

sjd78 replied:
You should remove it.  If the SSO doesn't work, a modal is already displayed with an awkward message to handle that scenario.

Right now, on the first run of the app, you'll see both the modal and the toast at the same time.  It isn't necessary.

sgratch commented on src/sagas/index.js:
```diff
@@ -183,6 +187,42 @@ export function* fetchByPage () {
+    } else if (selectedConsole === NO_DEFAULT_CONSOLE) {
+      yield put(addUserMessage({
+        type: 'error',
+        message: msg.consoleDefaultNotAvailable({ vmName }),
```
What about the defaultConsole as an input?

sgratch reviewed (changes requested):
@bamsalem 
- It crashes when trying to open a console for a headless VM:
![image](https://user-images.githubusercontent.com/18169498/108046705-9c2e3a00-704d-11eb-9679-25e992feba68.png)

- please fix PR's first comment based on previous review
- check inline comments

bamsalem commented:
@sgratch 
> @bamsalem
> 
> * It crashes when trying to open a console for a headless VM:
>   ![image](https://user-images.githubusercontent.com/18169498/108046705-9c2e3a00-704d-11eb-9679-25e992feba68.png)

It have been fixed, I didn't checked enough one of the suggested changes.

> * please fix PR's first comment based on previous review

Alredy Done.

> * check inline comments

I replayed to the relevant comments.

sjd78 commented on src/components/VmActions/AutoSelectConsole.js:
```diff
@@ -0,0 +1,104 @@
+    defaultConsole: vm.get('defaultConsole'),
+    isNoVnc: state.config.get('defaultVncMode') === 'NoVnc',
+  }),
+  (dispatch) => ({
+    fetchConsoles: (vm) => {
+      dispatch(downloadVmConsoles({ vm }))
+    },
+  })
```
```suggestion
  (dispatch, { vm }) => ({
    fetchConsoles: () => dispatch(downloadVmConsoles({ vm }),
  })
```

sjd78 commented on src/components/VmActions/AutoSelectConsole.js:
```diff
@@ -0,0 +1,104 @@
+    if (show) {
+      setDispatchStatus(SHOULD_DISPATCH)
+      fetchConsoles(vm)
+    }
+  }, [show, fetchConsoles])
+
+  useEffect(() => {
+    if (show && dispatchStatus === SHOULD_DISPATCH) {
+      if (!defaultConsole || defaultConsole === 'undefined') {
+        return
+      }
+
+      if (DEFAULT_CONSOLE_PROTOCOL_ERRORS_LIST.includes(defaultConsole)) {
+        handleClose()
+      } else {
+        setDispatchStatus(DISPATCHED)
+      }
+    }
+  }, [defaultConsole, dispatchStatus, show, handleClose])
+
+  const isProtocolSelected = show && dispatchStatus === DISPATCHED
+  return (
+    <div>
+      {isProtocolSelected &&
+      <ConsoleConfirmationModal
+        consoleId={defaultConsole.id}
+        vm={vm}
+        show
+        isNoVNC={defaultConsole.protocol === 'vnc' ? isNoVnc : false}
+        onClose={() => {
+          if (dispatchStatus === DISPATCHED) {
+            handleClose()
+          }
+        }}
+        id={id}
+      />
+      }
+    </div>
+  )
+}
```
I think you can get rid of component state use completely by just doing:

```jsx
}) => {
  useEffect(() => {
    if (show && defaultConsole === 'undefined') {
      fetchConsoles()
    }
  }, [show, defaultConsole])

  if (!show || defaultConsole === 'undefined' || defaultConsole === EMPTY_CONSOLES_LIST) {
    return null
  }

  return (
    <ConsoleConfirmationModal
      show
      vm={vm}
      consoleId={defaultConsole.id}
      isNoVNC={defaultConsole.protocol === 'vnc' ? isNoVnc : false}
      onClose={onClose}
      id={id}
    />
  )
}
```

Note 1: If the console is already fetched for the VM, it won't be fetched again.

Note 2: The `defaultConsole === 'undefined'` value is set in the base VM transform.  You can change it to whatever constant makes sense.  

In the case of a headless VM, clicking the action button will set `show=true` and then with `defaultConsole === 'undefined'` will hit the effect and call the fetch saga.  But, the saga would then set `deafulConsole` to `EMPTY_CONSOLES_LIST`.  The fetch won't be triggered again.  Only an error toast on the first click will be displayed.  Subsequent clicks will have no effect.

Perhaps if a VM is known to be headless, the console button can be removed from the action card or otherwise disabled to avoid confusion.  @sgratch, @rszwajko - Thoughts on multiple clicks to the console button for a headless VM?

bamsalem replied:
@sjd78
I got rid of the component's state but I made few small changes to integrate between the components because `ConsoleConfirmationModal` is class component and `AutoSelectConsole` is functional ConsoleConfirmationModal sometimes triggers the onClose with no reason and then the console wasn't fetched as expected.

sgratch replied:
@sjd78 @bamsalem 

> Perhaps if a VM is known to be headless, the console button can be removed from the action card or otherwise disabled to avoid confusion. @sgratch, @rszwajko - Thoughts on multiple clicks to the console button for a headless VM?

I think it might be confusing for the user if the second click will disable the console button till next hard refresh or console configuration changes on backend. I think it's better to leave it as is to keep the ui simple and persistent.

rszwajko replied:
> I think you can get rid of component state use completely

yep. This is the right direction.

> Perhaps if a VM is known to be headless, the console button can be removed from the action card or otherwise disabled to avoid confusion. @sgratch, @rszwajko - Thoughts on multiple clicks to the console button for a headless VM?

I would disable the action (make the button grey) to clearly indicate that this option is not available. Active buttons encourage users to click & try them i.e. users may try it few times hoping it will finally work.

bamsalem replied:
@sjd78 @rszwajko 
![image](https://user-images.githubusercontent.com/64131213/108827491-597cdc80-75ce-11eb-964c-eaa8bb524be4.png)

>I would disable the action (make the button grey) to clearly indicate that this option is not available. Active buttons encourage users to click & try them i.e. users may try it few times hoping it will finally work.

We can't disable the button without disabling the dropdown menu (with the current component, PF component).
and I'm not sure it's worth to create our own component for that, but we must add another notifier, maybe keep showing the error on every click...
WDYT?

sgratch replied:
@bamsalem for headless:
I suggest reverting back for keeping the headless error per every click.
 We can change the menu items order so that "console" menu item will not be the default till next refresh (the default/first item will be "Remote Desktop" for win based vms and "Suspend" for others).

bamsalem replied:
> @bamsalem for headless:
> I suggest reverting back for keeping the headless error per every click.

Done.

> We can change the menu items order so that "console" menu item will not be the default till next refresh (the default/first item will be "Remote Desktop" for win based vms and "Suspend" for others).

Regarding changing the order, it's more complex than it looks, because it's a PF component when we change the order it's change the confirmation modal and than it display non related notification, for example "Are you sure you want to Suspend the VM?", to handle that we need to implement our own component.

sgratch replied:
@bamsalem so we can go back to original behavior of just hidden/dismiss the "console" menu item till next refresh.

sjd78 reviewed (changes requested):
It's close.  `AutoSelectConsole` can be simpler.

sgratch commented on src/sagas/index.js:
```diff
@@ -183,6 +187,41 @@ export function* fetchByPage () {
+    // select the console to use based on availability and app/user configs
+    const selectedConsole =
+      consoles.length === 0 ? EMPTY_CONSOLES_LIST
+        : consoles.length === 1 ? consoles[0]
```
@sjd78 I think that we should expect the supported only console to be used in case only one protocol is supported, regardless to default console. The default console is only for cases in which all options are supported and we need to auto choose for the user.

bamsalem replied:
@sgratch @sjd78 
I changed the code back to the prev version (now the system will use the only console available).

sgratch commented:
@bamsalem please fix the TypeError:
![image](https://user-images.githubusercontent.com/18169498/109159952-e1edaf80-777d-11eb-823c-1997f45f90e2.png)

bamsalem commented:
> @bamsalem please fix the TypeError:
> ![image](https://user-images.githubusercontent.com/18169498/109159952-e1edaf80-777d-11eb-823c-1997f45f90e2.png)

@sgratch
I tried to reproduce the error but I couldn't, can you provide more details/steps to reproduce please?
BTW I'm not sure but it seems like the issue is not related to this fix, maybe we should deal with that in separate PR...

sgratch commented:
@bamsalem 
> I tried to reproduce the error but I couldn't, can you provide more details/steps to reproduce please?
> BTW I'm not sure but it seems like the issue is not related to this fix, maybe we should deal with that in a separate PR...

This error generated for me when app was up but it's not reproduced anymore.

sjd78 commented on src/components/VmActions/AutoSelectConsole.js:
```diff
@@ -0,0 +1,76 @@
+    if (show && DEFAULT_CONSOLE_PROTOCOL_ERRORS_LIST.includes(defaultConsole)) {
+      onClose()
+    }
+  }, [defaultConsole, show, onClose])
```
Why is `onClose` on the deps list?

sjd78 reviewed (changes requested):
Tests ok on my dev environment, but I have concerns about the deps list on AutoSelectConsole's `useEffect` hooks.

sjd78 commented:
> @bamsalem
> 
> > I tried to reproduce the error but I couldn't, can you provide more details/steps to reproduce please?
> > BTW I'm not sure but it seems like the issue is not related to this fix, maybe we should deal with that in a separate PR...
> 
> This error generated for me when app was up but it's not reproduced anymore.

line 444 of `transform.js` indicates that a Pool did not have a `vm` property returned from the rest api.  I've not seen that before - the user would have to have access to either (1) an empty pool, or (2) a pool that is visible to the user but ALL vms are not.  There may be other scenarios there, but they all seem very unlikely.

sgratch reviewed (approved):
LGTM except 2 issues to consider as follow up issues:
1. https://github.com/oVirt/ovirt-web-ui/pull/1342#discussion_r587488486 - consider removing the console menu option for headless - this can be implemented separately. 

2. Please note that If the consoles type list is already fetched for the VM, it is currently not fetched again till next refresh.
Until now, we didn't call fetchConsoles per every console click, but the user could see which types are supported and choose which to open on dashboard view.  If consoles can't be displayed/chosen on dashboard, at least fetching the consoles can be up to date when user tries to open one. It might add a delay for console opening though...
This should be considered again when implementing console settings.

sjd78 reviewed (approved):
LGTM
